pysteps/nowcasts/linda.py

```diff
@@ -0,0 +1,1359 @@
+        precip_fields.shape[1],
+        precip_fields.shape[2],
+        localization_window_radius,
+    )
+
+    kernels_1 = _estimate_convol_params(
+        precip_fields_lagr_diff[-2],
+        precip_fields_lagr_diff[-1],
+        convol_weights,
+        mask_adv,
+        kernel_type=kernel_type,
+        num_workers=num_workers,
+    )
+    fct_gen["kernels_1"] = kernels_1
+
+    if measure_time:
+        print("{:.2f} seconds.".format(time.time() - starttime))
+    else:
+        print("done.")
+
+    # compute convolved difference fields
+    precip_fields_lagr_diff_c = precip_fields_lagr_diff[:-1].copy()
+    for i in range(precip_fields_lagr_diff_c.shape[0]):
+        for j in range(ari_order - i):
+            precip_fields_lagr_diff_c[i] = _composite_convolution(
+                precip_fields_lagr_diff_c[i],
+                kernels_1,
+                interp_weights,
+            )
+
+    print("Estimating the ARI(p,1) parameters... ", end="", flush=True)
+
+    if measure_time:
+        starttime = time.time()
+
+    # estimate ARI(p,1) parameters
+    weights = _compute_window_weights(
+        feature_coords,
+        precip_fields.shape[1],
+        precip_fields.shape[2],
+        localization_window_radius,
+    )
+
+    if ari_order == 1:
+        psi = _estimate_ar1_params(
+            precip_fields_lagr_diff_c[-1],
+            precip_fields_lagr_diff[-1],
+            weights,
+            interp_weights,
+            num_workers=num_workers,
+        )
+    else:
+        psi = _estimate_ar2_params(
+            precip_fields_lagr_diff_c[-2:],
+            precip_fields_lagr_diff[-1],
+            weights,
+            interp_weights,
+            num_workers=num_workers,
+        )
+    fct_gen["psi"] = psi
+
+    if measure_time:
+        print("{:.2f} seconds.".format(time.time() - starttime))
+    else:
+        print("done.")
+
+    # apply the ARI(p,1) model and integrate the differences
+    precip_fields_lagr_diff_c = _iterate_ar_model(precip_fields_lagr_diff_c, psi)
+    precip_fct = precip_fields_lagr[-2] + precip_fields_lagr_diff_c[-1]
+    precip_fct[precip_fct < 0.0] = 0.0
+
+    print("Estimating the second convolution kernel... ", end="", flush=True)
+
+    if measure_time:
+        starttime = time.time()
+
+    # estimate the second convolution kernels based on the forecast field
+    # computed above
+    kernels_2 = _estimate_convol_params(
+        precip_fct,
+        precip_fields[-1],
+        convol_weights,
+        mask_adv,
+        kernel_type=kernel_type,
+        num_workers=num_workers,
+    )
+    fct_gen["kernels_2"] = kernels_2
+
+    if measure_time:
+        print("{:.2f} seconds.".format(time.time() - starttime))
+    else:
+        print("done.")
+
+    if measure_time:
+        return fct_gen, precip_fields_lagr_diff, time.time() - starttime_init
+    else:
+        return fct_gen, precip_fields_lagr_diff
+
+
+# compute "masked" convolution where non-finite values are ignored
```

Comment:
I thought that it's not needed because these are private methods that are not visible outside the module. Is that OK?

Reply:
mmh i see, true, but I would anyway change it for consistency and also to maintain good practiced throughout (but I agree that here it would not be a problem per se)